### PR TITLE
Build system rewrite

### PR DESCRIPTION
**Pull request by tact1m4n3**
## Build system rewrite

If anyone has any other ideas of things to add to the new build system (even stuff that was there before and I missed it), please leave a comment in this PR and I will add it to the list. 

### Tasks to complete
- [x] Initial version of the build system done.
- [x] Better way for targets to select a cpu/arch
- [x] Patch elf
- [x] Build formats
- [x] Better ux for port writing
- [ ] Change all ports to work with the new build system (wip)
    - [ ] rp2xxx: find a way to build bootrom and also expose it so as to allow user targets to reference it
    - [x] lpc
    - [x] stm32
    - [x] nrf5x
    - [x] gd32
    - [x] atsam
    - [x] avr
    - [ ] esp
- [ ] Change all examples to work with the new build system
    - [x] rp2xxx
    - [x] lpc
    - [x] stm32
    - [x] nrf5x
    - [x] gd32
    - [x] atsam
    - [x] avr
    - [ ] esp
- [x] Final touches
    - [x] Add the remaining options from the previous build system
    - [x] Add the `package` command back to the root `build.zig`

### CODE REVIEW

File: port/nxp/lpc/build.zig [hunk 1]
```diff
@@ -1,67 +1,72 @@
...
+/// Patch an ELF file to add a checksum over the first 8 words so the
 /// cpu will properly boot.
-fn postprocess(b: *std.Build, input: std.Build.LazyPath) std.Build.LazyPath {
+fn patch_elf(b: *std.Build, input: std.Build.LazyPath) std.Build.LazyPath {
     const patchelf = b.addExecutable(.{
         .name = "lpc176x5x-patchelf",
-        .root_source_file = path("/src/tools/patchelf.zig"),
+        .root_source_file = b.path("src/tools/patchelf.zig"),
         .target = b.host,
     });
```
Comment (ikskuh) [inline port/nxp/lpc/build.zig:67]: The `add_executable` should happen in `fn build()` and the call to `patch_elf` should receive the `std.Build.Dependency` to this module, so you can use `dep.artifact("lpc176x5x-patchelf")` instead of creating a new executable for each patched ELF file

File: port/nxp/lpc/build.zig [hunk 2]
```diff
@@ -1,67 +1,72 @@
...
 
-/// Post-processes an ELF file to add a checksum over the first 8 words so the
+    internals.submit_target(chips.lpc176x5x, .{
+        .chip = lpc176x5x_chip,
+        .hal = lpc176x5x_hal,
+        .preferred_binary_format = .elf,
+        .patch_elf = .{
+            .b = b,
+            .func = patch_elf,
+        },
```
Comment (ikskuh) [inline port/nxp/lpc/build.zig:45]: Does it make sense that the `patch_elf` holds a reference to the constructing builder?

File: build-internals/build.zig [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+
+    pub fn init(name: []const u8) TargetAlias {
+        return .{ .name = name };
+    }
+};
+
+pub const Chip = struct {
+    b: *Build,
+
+    name: []const u8,
```
Comment (ikskuh) [inline build-internals/build.zig:58]: This can be internally set in submit_target from the target alias

File: build-internals/build.zig [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+    pub const Custom = struct {
+        /// The standard extension of the format.
+        extension: []const u8,
+
+        /// A function that will convert a given `elf` file into the custom output format.
+        ///
+        /// The `*Custom` format is passed so contextual information can be obtained by using
+        /// `@fieldParentPtr` to provide access to tooling.
+        convert: *const fn (*Custom, elf: Build.LazyPath) Build.LazyPath,
+    };
```
Comment (ikskuh) [inline build-internals/build.zig:225]: `Custom` needs the providing `*std.Dependency` or `*std.Build` otherwise we can't allocate memory or access custom tooling (via `b.dependency` or `b.artifact`)

**Review by ikskuh — CHANGES_REQUESTED** [review]

File: build-internals/build.zig.zon [hunk 5]
```diff
@@ -1,8 +1,8 @@
 .{
-    .name = "build/definitions",
+    .name = "microzig/build-internals",
```
Comment (mattnite) [inline build-internals/build.zig.zon:2]: `build-internals`

File: tools/uf2/build.zig [hunk 6]
```diff
@@ -1,34 +1,5 @@
 const std = @import("std");
-const Dependency = std.Build.Dependency;
```
Comment (mattnite) [inline tools/uf2/build.zig:2]: Why is all of this being deleted? This is unrelated to the purpose of this PR.

File: tools/regz/build.zig [hunk 7]
```diff
@@ -1,8 +1,5 @@
 const std = @import("std");
 const Build = std.Build;
-const Compile = Build.Compile;
```
Comment (mattnite) [inline tools/regz/build.zig:3]: I get that you're cleaning up, please remove this as this patch is very large. Do it in a different patch.

File: build.zig [hunk 8]
```diff
@@ -1,146 +1,566 @@
...
+            .root_source_file = switch (cpu) {
+                .avr5 => b.path("cpus/avr5.zig"),
+                .cortex_m => b.path("cpus/cortex_m.zig"),
+                .riscv32 => b.path("cpus/riscv32.zig"),
+            },
+        });
+    }
+};
+
+pub inline fn customLazyImport(
```
Comment (mattnite) [inline build.zig:433]: snake_case

File: build.zig [hunk 9]
```diff
@@ -1,146 +1,566 @@
...
+                pkg.build_zig
+            else
+                @compileError("dependency '" ++ dep_name ++ "' does not have a build.zig");
+        }
+    }
+
+    comptime unreachable; // Bad @dependencies source
+}
+
+inline fn customFindImportPkgHashOrFatal(comptime dep_name: []const u8) []const u8 {
```
Comment (mattnite) [inline build.zig:457]: snake_case

File: build.zig [hunk 10]
```diff
@@ -1,146 +1,566 @@
...
 
-    const parts_db_str = try std.json.stringifyAlloc(b.allocator, parts_db, .{
-        .whitespace = .indent_4,
-    });
-    const write_file_step = b.addWriteFiles();
-    return write_file_step.add("parts-db.json", parts_db_str);
+    @panic("dependency not found");
 }
+
+// fn init_cpu_map(mz: *MicroZig) void {
```
Comment (mattnite) [inline build.zig:474]: Either include or delete this commented out code

File: build.zig.zon [hunk 11]
```diff
@@ -1,46 +1,27 @@
...
-        .@"tools/uf2" = .{ .path = "tools/uf2" },
-        .@"port/nordic/nrf5x" = .{ .path = "port/nordic/nrf5x" },
-        .@"port/nxp/lpc" = .{ .path = "port/nxp/lpc" },
-        .@"port/microchip/atsam" = .{ .path = "port/microchip/atsam" },
-        .@"port/microchip/avr" = .{ .path = "port/microchip/avr" },
-        .@"port/gigadevice/gd32" = .{ .path = "port/gigadevice/gd32" },
-        .@"port/stmicro/stm32" = .{ .path = "port/stmicro/stm32" },
-        .@"port/espressif/esp" = .{ .path = "port/espressif/esp" },
-        .@"port/raspberrypi/rp2xxx" = .{ .path = "port/raspberrypi/rp2xxx" },
+        .@"microzig/build-internals" = .{ .path = "build-internals" },
```
Comment (mattnite) [inline build.zig.zon:5]: dont' prefix these packages with microzig

File: build.zig.zon [hunk 12]
```diff
@@ -1,46 +1,27 @@
...
 
-        // examples so that we can build them all in one go
```
Comment (mattnite) [inline build.zig.zon:21]: If you're going to put all the example packages under one, please add the dependency here.

File: core/build.zig.zon [hunk 13]
```diff
@@ -1,15 +0,0 @@
-.{
-    .name = "core",
```
Comment (mattnite) [inline core/build.zig.zon:2]: Why did you remove this package? Do you understand that you've broken the packaging system?

**Review by mattnite — CHANGES_REQUESTED** [review]
If each port is now a type, is it possible to add out of source ports? If we can't is this a direction we want to take the project?

This patch is big and changes a lot, please minimize it as much as you can. I don't understand why altering the structure of the core package is useful to this task. Same thing with a lot of your deletions, if you're going to clean up please put it in a different PR.

**Comment by tact1m4n3** [comment]
> If each port is now a type, is it possible to add out of source ports? If we can't is this a direction we want to take the project?

Sure. This is how it would look if you followed the same design as the ports in the repo (though this is not required).
```zig
pub fn build(b: *std.Build) void {
    const optimize = b.standardOptimizeOption(.{});

    const mz_dep = b.dependency("microzig", .{});
    const mz = MicroZig.init(b, mz_dep);

    const port_dep = b.dependency("port", .{});
    const port = @import("port").init(port_dep);

    const fw = mz.add_firmware(.{
        .name = "test",
        .root_source_file = b.path("src/main.zig"),
        .target = port.whatever,
        .optimize = optimize,
    });
    mz.install_firmware(fw, .{});
}
```

> I don't understand why altering the structure of the core package is useful to this task.

I did this because since the root build.zig contains the build system, it seemed logical to make `core/microzig.zig` a part of this package (since it is the one that merges all the moving parts we have to add all the necessary imports, which can be done even if the module belongs to a different package, but it makes sense that the package that creates the module also adds all the imports). The same goes for `core/start.zig` or any other modules (chip, hal, board). They are constructed by the build system.
Another reason is that core can't be used on it's own (at least in it's current form), it would only ever be used by the microzig build system. There are some definitions that I think can be placed in another package (core) that contains `Mmio`, etc. This would allow other packages to access those definitions.

I apologize for the excessive amount of changes and for making you navigate the long long diff :). I guess my lack of experience with open source really shows here. I went through several iterations of the code (without starting fresh) and I changed a lot of things I didn't write in the process. If there are two many changes, I can make a cleaner pr.

**Comment by tact1m4n3** [comment]
I agree that this is not a pr, but more of a playground of ideas and it contains stuff not necessarily related to the purpose of the original pr. If it's ok with you, I'll make a new pr with only the only changes being to the build system or force push in this one. There is also a bug in the stm32 chip generation (fixed here) for which I'll make a different pr.

File: core/build.zig.zon (hunk 13 above)
Comment (tact1m4n3) [reply core/build.zig.zon:2]: My intention was to include core (microzig, start and cpus) directly into the root package (didn't update each package's files list, yet).

Comment (tact1m4n3) [reply core/build.zig.zon:2]: I added it back

File: build.zig.zon (hunk 12 above)
Comment (tact1m4n3) [reply build.zig.zon:21]: I tried this and it gives me a strange error. Otherwise, building all the examples with `zig build` works fine.
```
/home/toto/dev/zig-embedded/microzig/examples/build.zig:1:1: error: file exists in multiple modules
/home/toto/dev/zig-embedded/microzig/examples/build.zig:1:1: note: root of module root.@dependencies.122004ac89e8f0168f24759576ced3d80b47ccd0c0fe4d923e83794d8a813b026932
/home/toto/dev/zig-embedded/microzig/examples/build.zig:1:1: note: root of module root.@build
/home/toto/dev/zig-embedded/microzig/examples/../examples/raspberrypi/rp2xxx/build.zig:1:1: error: file exists in multiple modules
/home/toto/dev/zig-embedded/microzig/examples/../examples/raspberrypi/rp2xxx/build.zig:1:1: note: root of module root.@dependencies.122036a1b1203ff16a15cdf6e496f567225abc589ef4295bf109cca8758b3c27e304
/home/toto/dev/zig-embedded/microzig/examples/../examples/raspberrypi/rp2xxx/build.zig:1:1: note: root of module root.@dependencies.122016b96a0fddd1a5ea10a2f969decf63be550014831b01cc906b441c300f15c801
/home/toto/dev/zig-embedded/microzig/examples/nxp/lpc/build.zig:1:1: error: file exists in multiple modules
/home/toto/dev/zig-embedded/microzig/examples/nxp/lpc/build.zig:1:1: note: root of module root.@dependencies.12208201167b38fd4b7c7b676e2a973d198ec63f809045b8a55654f72d5dbf4c9e21
/home/toto/dev/zig-embedded/microzig/examples/nxp/lpc/build.zig:1:1: note: root of module root.@dependencies.122020eda6f6de553d128be12ae72e608c7f060f091ee0920f2348da5a0fafbc6dc3
/home/toto/dev/zig-embedded/microzig/examples/../examples/stmicro/stm32/build.zig:1:1: error: file exists in multiple modules
/home/toto/dev/zig-embedded/microzig/examples/../examples/stmicro/stm32/build.zig:1:1: note: root of module root.@dependencies.12207ea8a92f85718d537bfc361d17e2010c2c612598825adf035e2bd23e57df342d
/home/toto/dev/zig-embedded/microzig/examples/../examples/stmicro/stm32/build.zig:1:1: note: root of module root.@dependencies.1220461576a7eef1ad0195ef9b32c632712429bb6e6cc6f32a6757eaf66995c720ab
```

Comment (mattnite) [reply build.zig.zon:21]: That's fair then, I'm not sure what exactly the issue is but we can leave it for now

**Comment by mattnite** [comment]
@tact1m4n3 Thank you for the example, I really like how this is turning out. No worries on being new to open source, you've done some great work here and I'm happy to help get this into the main branch. As a bit of extra information, we do squash commits here, so in the final history every PR will result in a single commit. This makes it so that one feature, bug fix, etc is reflected by every commit, so when you look at a patch, it's a meaningful addition to the codebase.

You're making some big changes to the build system so it makes sense that it'll touch a lot of files, but if you can please make those other, PRs that you mentioned!

**Comment by tact1m4n3** [comment]
Just fixed the build error and merged the branch with upstream/main.

> if you can please make those other, PRs that you mentioned!

Can you clarify please to what PRs are you referring to?

> There is also a bug in the stm32 chip generation (fixed here) for which I'll make a different pr.

I've already submitted this one. #282

**Comment by tact1m4n3** [comment]
> I've already submitted this one. #282

I removed the changes addressed by #282 from this PR. I think it would be better to merge this one first (there are conflicts between them) because then it would be clearer what PR patched what.
